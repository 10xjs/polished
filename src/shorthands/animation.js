// @flow
/** @module Shorthands */

/**
 * Shorthand for easily setting the animation property
 * @static
 * @param {...(string|number)} [arguments] - Up to eight animation properties
 * @return {Object} CSS as JS.
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...animation('rotate', '1s', 'ease-in-out')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${animation('rotate', '1s', 'ease-in-out')}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'animation': 'rotate 1s ease-in-out'
 * }
 */

function animation(...args: Array<string|number>) {
  if (args.length > 8) {
    throw new Error('The animation shorthand only takes 8 arguments. See the specification for more information: http://mdn.io/animation')
  }

  return {
    'animation': args.join(' '),
  }
}

export default animation
