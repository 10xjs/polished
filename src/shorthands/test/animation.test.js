// @flow
import animation from '../animation'

describe('animation', function() {
  it('should pass first eight arguments to the CSS', function() {
    expect({ ...animation('rotate', '1s', 'ease-in-out', '0.5s', 5, 'reverse', 'forwards', 'paused') }).toMatchSnapshot()
  })

  it('should be fine with less than eight arguments', function() {
    expect({ ...animation('rotate', '1s', 'ease-in-out') }).toMatchSnapshot()
  })

  it('should throw an error if more than eight arguments are supplied', function() {
    expect(function () { animation('one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'oops') }).toThrow()
  })
})
