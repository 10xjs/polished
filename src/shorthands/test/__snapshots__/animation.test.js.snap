exports[`animation should be fine with less than eight arguments 1`] = `
Object {
  "animation": "rotate 1s ease-in-out",
}
`;

exports[`animation should pass first eight arguments to the CSS 1`] = `
Object {
  "animation": "rotate 1s ease-in-out 0.5s 5 reverse forwards paused",
}
`;
